2_process.smk

```diff
@@ -0,0 +1,155 @@
+import os
+import numpy as np
+import pandas as pd
+
+configfile: "2_process/process_config.yaml"
+
+
+def mntoha_lake_sequence_files(wildcards):
+    """
+    List all MNTOHA lake sequence .npy files for training and testing.
+
+    :param wildcards: Snakemake wildcards (empty).
+    :returns: List of lake training/testing sequence files.
+
+    """
+    # make this function dependent on fetch_all
+    # needed because lake_metadata.csv is used to determine lake_sequence_files
+    lake_metadata_file = checkpoints.fetch_mntoha_metadata.get(**wildcards).output[0]
```

Comment:
Just to make sure I'm interpreting this correctly --- here, we're checking if the `fetch_mntoha_metadata` checkpoint has been executed for _everything_ that falls under that checkpoint rule, because of that `**wildcards` notation, right, even though `wildcards` is empty in this case, as you note in the function documentation. And if `wildcards` weren't empty, and there were one named `category`, we could access the output files associated with a category named 'minnesota' using `checkpoints.fetch_mntoha_metadata.get(category="minnesota").output`, right? For reference, I'm reading [these docs](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html) to try to piece it all together.

Reply:
Or are we just checking that the checkpoint has been executed for that first (and in this case, only) output file `1_fetch/out/lake_metadata.csv`? But, in theory if there were multiple wildcards, it would check that the first output file had been generated for each of them?

Sorry for the snakemake Qs - I know that's not the point of my review -- just trying to get oriented to the pipeline.

Reply:
The `**wildcards` notation means that whatever wildcards were passed as kwargs into `mntoha_lake_sequence_files` are getting passed on through as kwargs to the checkpoint `fetch_mntoha_metadata`. You're right that `wildcards` is empty here. See `dynamic_filenames` for a case where `wildcards` isn't empty. Yes, you could pass specific values for wildcards to a checkpoint as you wrote. Again, that happens in `dynamic_filenames`.

Reply:
In `dynamic_filenames`, even though `wildcards` isn't empty, it's unused by the checkpoint `fetch_mntoha_metadata` because there's no wildcards defined in that checkpoint.

Reply:
Ok, cool, that make sense. Thanks for the explanation!